src/StepFunctionGraph/Graph.tsx

```diff
@@ -0,0 +1,423 @@
+            setZoom(window.devicePixelRatio);
+        }
+    };
+
+    // Gets better coordinates for the mouse inside the canvas
+    const getMousePos = (canvas: any, evt: any) => {
+        const rect = canvas.getBoundingClientRect();
+        return {
+            x: (evt.clientX - rect.left) / (rect.right - rect.left) * canvas.width,
+            y: (evt.clientY - rect.top) / (rect.bottom - rect.top) * canvas.height
+        };
+    };
+
+    const getClickHandler = (drawn: Map<number, NodeDimensions>, ctx: CanvasRenderingContext2D, canvasRef: any) => {
+        // This function checks for any clicks inside the bounds of any drawn rectangles
+        return function nodeClickHandler(e: any) {
+            const { x: offsetX, y: offsetY } = getMousePos(canvasRef, e);
+            const rects = Array.from(drawn.values())
+                .filter(node =>
+                    node.nodeType !== WorkFlowType.START &&
+                    node.nodeType !== WorkFlowType.END
+                );
+            const collision: NodeDimensions | null = collides(
+                rects,
+                offsetX,
+                offsetY,
+                ctx
+            );
+
+            handleNodeClick(collision);
+        };
+    };
+
+    // Recalculate the center of the graph based on zoom and current canvas dimensions
+    const calculateCenter = (): void => {
+        const ctx = canvasContainer?.current?.getContext('2d');
+
+        if (ctx) {
+            const center = {
+                x: -(zoom - 1) * ctx.canvas.width / 2,
+                y: -(zoom - 1) * (ctx.canvas.height / 2)
+            };
+
+            globalOffset.offset.x = -center.x / 2;
+            globalOffset.offset.y = -center.y / 2;
+
+            globalOffset.defaultOffset.x = -center.x / 2;
+            globalOffset.defaultOffset.y = -center.y / 2;
+
+            pan.offset.x = -center.x / 2;
+            pan.offset.y = -center.y / 2;
+        }
+    };
+
+    // Pans the canvas to a given node. Parallel nodes are calculated from the side and not center.
+    const panToNode = (node: NodeDimensions | null): void => {
+        const ctx = canvasContainer?.current?.getContext('2d');
+        if (ctx && node?.x) {
+            const isParallel = node.nodeType === WorkFlowType.PARALLEL;
+            const xOffset = isParallel ? node.x + node.width / 2 : node.x;
+
+            pan.offset.x = (globalOffset.defaultOffset.x * 2) - xOffset;
+            pan.offset.y = ((globalOffset.defaultOffset.y * 2) - (node.y));
+            globalOffset.offset.x = (globalOffset.defaultOffset.x * 2) - xOffset;
+            globalOffset.offset.y = ((globalOffset.defaultOffset.y * 2) - (node.y));
+        }
+    };
+
+    // Construct a Digraph to store a representation of the graph to draw
+    const graph: Digraph = useMemo(() => new Digraph(), [json]);
+
+    // Parse the JSON and return a complete graph
+    // This States field check can be removed once we fit the JSON to an interface
```

Comment:
Do you think this should be marked with a TODO or is that not necessary?

Reply:
I'll just remove the comment about it. I'm not sure an interface will be written for it.